# Slicing and Indexing DataFrames

Indexes are supercharged row and column names. Learn how they can be combined with slicing for powerful DataFrame subsetting.


## Explicit Indexes


### .columns and .index
```python
dogs.columns
```

```python
dogs.index
```

dot-columns contains an Index object of column names, and dot-index contains an Index object of row numbers.

### Setting a column as the index

```python
dog_ind = dogs.set_index("name")
```

### Removing an index

```python
dogs_index.reset_index()
```

### Dropping an index
```python
dogs_ind.reset_index(drop=True)
```

reset_index has a drop argument that allows you to discard an index. Here, setting drop to True entirely removes the dog names.

### Indexes make subsetting simpler
```python
dogs[dogs["name"].isin["Bella", "Stella"]]
```

```python
dogs_ind.loc[["Bella", "Stella"]]
```

DataFrames have a subsetting method called "loc," which filters on index values. Here you simply pass the dog names to loc as a list. Both work same things


### Index values don't need to be unique
```python
dogs_ind2 = dogs.set_index("breed")
```

### Subsetting on duplicated index values
```python
dogs_ind2.loc["Labrador"]
```
if you subset on "Labrador" using loc, all the Labrador data is returned.


### Multi-level indexes a.k.a. hierarchical indexes
```python
dogs_ind3 = dogs.set_index(["breed", "color"])
```
There is an implication here that the inner level of index, in this case, color, is nested inside the outer level, breed.

### Subset the outer level with a list
```python
dogs_ind3.loc[["Labrador", "Chihuahua"]]
```

To take a subset of rows at the outer level index, you pass a list of index values to loc. Here, the list contains Labrador and Chihuahua, and the resulting subset contains all dogs from both breeds.

### Subset the outer level with a list
```python
dogs_ind3.loc[["Labrador", "Chihuahua"]]
```

### Subset inner levels with a list of tuples
```python
dogs_ind3.loc[[("Labrador", "Brown"), ("Chihuahua", "Taan")]]
```

To subset on inner levels, you need to pass a list of tuples. Here, the first tuple specifies Labrador at the outer level and Brown at the inner level. 

### Sorting by Index Values
```python
dogs_ind3.sort_index()
```
 it sorts all index levels from outer to inner, in ascending order.


### Controlling sort_index

```python
dogs_ind3.sort_index(level=["color", "breed"], ascending=[True, False])
```

### Now You have two problems
- Index values are just data
- Indexes violate `"tidy data"` principles
- You need to learn two syntaxes
- 

### Setting and removing indexes

pandas allows you to designate columns as an *index* . This enables cleaner code when taking subsets (as well as providing more efficient lookup under some circumstances).

In this chapter, you'll be exploring `temperatures`, a DataFrame of average temperatures in cities around the world. `pandas` is loaded as `pd`.

#### Instructions

* *Look at `temperatures`* .
* Set the index of `temperatures` to `"city"`, assigning to `temperatures_ind`.
* *Look at `temperatures_ind`. How is it different from `temperatures`?*
* Reset the index of `temperatures_ind`, keeping its contents.
* Reset the index of `temperatures_ind`, dropping its contents.


In [5]:
import  pandas as pd
temperatures = pd.read_csv("../../data/temperatures.csv")

In [6]:
# Look at temperatures
print(temperatures)

# Set the index of temperatures to city
temperatures_ind = temperatures. set_index("city")

# Look at temperatures_ind
print(temperatures_ind)

# Reset the temperatures_ind index, keeping its contents
print(temperatures_ind.reset_index())

# Reset the temperatures_ind index, dropping its contents
print(temperatures_ind.reset_index(drop=True))

       Unnamed: 0        date     city        country  avg_temp_c
0               0  2000-01-01  Abidjan  Côte D'Ivoire      27.293
1               1  2000-02-01  Abidjan  Côte D'Ivoire      27.685
2               2  2000-03-01  Abidjan  Côte D'Ivoire      29.061
3               3  2000-04-01  Abidjan  Côte D'Ivoire      28.162
4               4  2000-05-01  Abidjan  Côte D'Ivoire      27.547
...           ...         ...      ...            ...         ...
16495       16495  2013-05-01     Xian          China      18.979
16496       16496  2013-06-01     Xian          China      23.522
16497       16497  2013-07-01     Xian          China      25.251
16498       16498  2013-08-01     Xian          China      24.528
16499       16499  2013-09-01     Xian          China         NaN

[16500 rows x 5 columns]
         Unnamed: 0        date        country  avg_temp_c
city                                                      
Abidjan           0  2000-01-01  Côte D'Ivoire      27.293
Abidj

### Subsetting with .loc[]

The killer feature for indexes is `.loc[]`: a subsetting method that accepts index values. When you pass it a single argument, it will take a subset of rows.

The code for subsetting using `.loc[]` can be easier to read than standard square bracket subsetting, which can make your code less burdensome to maintain.

`pandas` is loaded as `pd`. `temperatures` and `temperatures_ind` are available; the latter is indexed by `city`.

#### Instructions

* Create a list called `cities` that contains "Moscow" and "Saint Petersburg".
* Use `[]` subsetting to filter `temperatures` for rows where the `city` column takes a value in the `cities` list.
* Use `.loc[]` subsetting to filter `temperatures_ind` for rows where the city is in the `cities` list.


In [8]:
cities = ["Moscow", "Saint Petersburg"]

# Subset temperatures using square brackets
print(temperatures[temperatures["city"].isin(cities)])

# Subset temperatures_ind using .loc[]
print(temperatures_ind.loc[cities])

       Unnamed: 0        date              city country  avg_temp_c
10725       10725  2000-01-01            Moscow  Russia      -7.313
10726       10726  2000-02-01            Moscow  Russia      -3.551
10727       10727  2000-03-01            Moscow  Russia      -1.661
10728       10728  2000-04-01            Moscow  Russia      10.096
10729       10729  2000-05-01            Moscow  Russia      10.357
...           ...         ...               ...     ...         ...
13360       13360  2013-05-01  Saint Petersburg  Russia      12.355
13361       13361  2013-06-01  Saint Petersburg  Russia      17.185
13362       13362  2013-07-01  Saint Petersburg  Russia      17.234
13363       13363  2013-08-01  Saint Petersburg  Russia      17.153
13364       13364  2013-09-01  Saint Petersburg  Russia         NaN

[330 rows x 5 columns]
                  Unnamed: 0        date country  avg_temp_c
city                                                        
Moscow                 10725  2000-01-

### Setting multi-level indexes

Indexes can also be made out of multiple columns, forming a *multi-level index* (sometimes called a *hierarchical index* ). There is a trade-off to using these.

The benefit is that multi-level indexes make it more natural to reason about nested categorical variables. For example, in a clinical trial, you might have control and treatment groups. Then each test subject belongs to one or another group, and we can say that a test subject is nested inside the treatment group. Similarly, in the temperature dataset, the city is located in the country, so we can say a city is nested inside the country.

The main downside is that the code for manipulating indexes is different from the code for manipulating columns, so you have to learn two syntaxes and keep track of how your data is represented.

`pandas` is loaded as `pd`. `temperatures` is available.

#### Instructions

**100 XP**

* Set the index of `temperatures` to the `"country"` and `"city"` columns, and assign this to `temperatures_ind`.
* Specify two country/city pairs to keep: `"Brazil"`/`"Rio De Janeiro"` and `"Pakistan"`/`"Lahore"`, assigning to `rows_to_keep`.
* Print and subset `temperatures_ind` for `rows_to_keep` using `.loc[]`.


In [10]:
# Index temperatures by country & city
temperatures_ind = temperatures.set_index(["country", "city"])

# List of tuples: Brazil, Rio De Janeiro & Pakistan, Lahore
rows_to_keep = [("Brazil", "Rio De Janeiro"), ("Pakistan", "Lahore")]

# Subset for rows to keep
print(temperatures_ind.loc[rows_to_keep])

                         Unnamed: 0        date  avg_temp_c
country  city                                              
Brazil   Rio De Janeiro       12540  2000-01-01      25.974
         Rio De Janeiro       12541  2000-02-01      26.699
         Rio De Janeiro       12542  2000-03-01      26.270
         Rio De Janeiro       12543  2000-04-01      25.750
         Rio De Janeiro       12544  2000-05-01      24.356
...                             ...         ...         ...
Pakistan Lahore                8575  2013-05-01      33.457
         Lahore                8576  2013-06-01      34.456
         Lahore                8577  2013-07-01      33.279
         Lahore                8578  2013-08-01      31.511
         Lahore                8579  2013-09-01         NaN

[330 rows x 3 columns]


### Sorting by index values

Previously, you changed the order of the rows in a DataFrame by calling `.sort_values()`. It's also useful to be able to sort by elements in the index. For this, you need to use `.sort_index()`.

`pandas` is loaded as `pd`. `temperatures_ind` has a multi-level index of `country` and `city`, and is available.

#### Instructions

* Sort `temperatures_ind` by the index values.
* Sort `temperatures_ind` by the index values at the `"city"` level.
* Sort `temperatures_ind` by ascending country then descending city.


In [12]:
# Sort temperatures_ind by index values
print(temperatures_ind.sort_index())

# Sort temperatures_ind by index values at the city level
print(temperatures_ind.sort_index(level="city"))

# Sort temperatures_ind by country then descending city
print(temperatures_ind.sort_index(level=["country", "city"], ascending=[True, False]))

                    Unnamed: 0        date  avg_temp_c
country     city                                      
Afghanistan Kabul         7260  2000-01-01       3.326
            Kabul         7261  2000-02-01       3.454
            Kabul         7262  2000-03-01       9.612
            Kabul         7263  2000-04-01      17.925
            Kabul         7264  2000-05-01      24.658
...                        ...         ...         ...
Zimbabwe    Harare        5605  2013-05-01      18.298
            Harare        5606  2013-06-01      17.020
            Harare        5607  2013-07-01      16.299
            Harare        5608  2013-08-01      19.232
            Harare        5609  2013-09-01         NaN

[16500 rows x 3 columns]
                       Unnamed: 0        date  avg_temp_c
country       city                                       
Côte D'Ivoire Abidjan           0  2000-01-01      27.293
              Abidjan           1  2000-02-01      27.685
              Abidjan      

## Slicing and subsetting with .loc and .iloc

### Slicing list

```python
breeds = ["Labrador", "Poodle", "Chow Chow", "Schnauzer", "Labrador", "Chihuahua", "St. Bernard"]
```

```python
breeds[2:5]
```
```text
['Chow Chow', 'Schnauzer', 'Labrador']
```

```python
breeds[:3]
```
```text
['Labrador', 'Poodle', 'Chow Chow']
```

```python
breeds[:]
```

```text
['Labrador','Poodle','Chow Chow','Schnauzer', 'Labrador','Chihuahua','St. Bernard']
```



### Sort the index before you slice
```python
dogs_srt = dogs.set_index(["breed", "color"]).sort_index()
print(dogs_srt)
```

### Slicing the outer index level
To slice rows at the outer level of an index, you call `loc`, passing the first and last values separated by a colon.

```python
dogs_srt.loc["Chow Chow":"Poodle"]
```
The final value `"Poodle"` is included

### Slicing the inner index levels badly
The same technique doesn't work on inner index levels.

It's important to understand the danger here. pandas doesn't throw an error to let you know that there is a problem, so be careful when coding.

```python
dogs_srt.loc["Tan":"Grey"]
```

### Slicing the inner index levels correctly

The correct approach to slicing at inner index levels is to pass the first and last positions as tuples.

```python
dogs_srt.loc[("Labrador", "Brown"):("Schnauzer", "Grey")]
```

### Slicing columns
The simplest case involves subsetting columns but keeping all rows. To do this, pass a colon as the `first argument` to `loc`. As with slicing lists, a `colon` by itself means `"keep everything."` The second argument takes `column names` as the `first and last` positions to slice on.

```python
dogs_srt.loc[:, "name":"height_cm"]
```



### Slice twice
You can slice on rows and columns at the same time: simply pass the appropriate slice to each argument.

```python
dogs_srt.loc[("Labrador", "Brown"):("Schnauzer", "Grey"), "name":"height_cm"]
```


### Dog days
An important use case of slicing is to subset DataFrames by a range of dates. To demonstrate this, let's set the date_of_birth column as the index and sort by this index.

```python
dogs = dogs.set_index("date_of_birth").sort_index()
```

### Slicing by dates
```python
# Get dogs with date_of_birth between 2014-08-25 and 2016-09-16
dogs.loc["2014-08-25":"2016-09-16"]
```

### Slicing by partial dates

```python
# Get dogs with date_of_birth between 2014-01-01 and 2016-12-31
dogs.loc["2014":"2016"]
```

### Subsetting by row/column number

You can also slice DataFrames by row or column number using the iloc method. This uses a similar syntax to slicing lists, except that there are two arguments: one for rows and one for columns. Notice that, like list slicing but unlike loc, the final values aren't included in the slice. In this case, the fifth row and fourth column aren't included.

```python
print(dogs.iloc[2:5, 1:4])
```

### Slicing index values

Slicing lets you select consecutive elements of an object using `first:last` syntax. DataFrames can be sliced by index values or by row/column number; we'll start with the first case. This involves slicing inside the `.loc[]` method.

Compared to slicing lists, there are a few things to remember.

* You can only slice an index if the index is sorted (using `.sort_index()`).
* To slice at the outer level, `first` and `last` can be strings.
* To slice at inner levels, `first` and `last` should be tuples.
* If you pass a single slice to `.loc[]`, it will slice the rows.

`pandas` is loaded as `pd`. `temperatures_ind` has country and city in the index, and is available.

#### Instructions

* Sort the index of `temperatures_ind`.
* Use slicing with `.loc[]` to get these subsets:
  * from Pakistan to Russia.
  * from Lahore to Moscow. (*This will return nonsense.* )
  * from Pakistan, Lahore to Russia, Moscow.


In [13]:
# Sort the index of temperatures_ind
temperatures_srt = temperatures_ind.sort_index()

# Subset rows from Pakistan to Russia
print(temperatures_srt.loc["Pakistan":"Russia"])

# Try to subset rows from Lahore to Moscow
print(temperatures_srt.loc["Lahore":"Moscow"])

# Subset rows from Pakistan, Lahore to Russia, Moscow
print(temperatures_srt.loc[("Pakistan", "Lahore"): ("Russia", "Moscow")])

                           Unnamed: 0        date  avg_temp_c
country  city                                                
Pakistan Faisalabad              4785  2000-01-01      12.792
         Faisalabad              4786  2000-02-01      14.339
         Faisalabad              4787  2000-03-01      20.309
         Faisalabad              4788  2000-04-01      29.072
         Faisalabad              4789  2000-05-01      34.845
...                               ...         ...         ...
Russia   Saint Petersburg       13360  2013-05-01      12.355
         Saint Petersburg       13361  2013-06-01      17.185
         Saint Petersburg       13362  2013-07-01      17.234
         Saint Petersburg       13363  2013-08-01      17.153
         Saint Petersburg       13364  2013-09-01         NaN

[1155 rows x 3 columns]
                    Unnamed: 0        date  avg_temp_c
country city                                          
Mexico  Mexico           10230  2000-01-01      12.694
    

### Slicing in both directions

You've seen slicing DataFrames by rows and by columns, but since DataFrames are two-dimensional objects, it is often natural to slice both dimensions at once. That is, by passing two arguments to `.loc[]`, you can subset by rows and columns in one go.

`pandas` is loaded as `pd`. `temperatures_srt` is indexed by country and city, has a sorted index, and is available.

#### Instructions

* Use `.loc[]` slicing to subset rows from India, Hyderabad to Iraq, Baghdad.
* Use `.loc[]` slicing to subset columns from `date` to `avg_temp_c`.
* Slice in both directions at once from Hyderabad to Baghdad, and `date` to `avg_temp_c`.


In [16]:
# Subset rows from India, Hyderabad to Iraq, Baghdad
print(temperatures_srt.loc[("India", "Hyderabad"): ("Iraq", "Baghdad")])

# Subset columns from date to avg_temp_c
print(temperatures_srt.loc[:, "date": "avg_temp_c"])

# Subset in both directions at once
print(temperatures_srt.loc[("India", "Hyderabad"): ("Iraq", "Baghdad"), "date": "avg_temp_c"])

                   Unnamed: 0        date  avg_temp_c
country city                                         
India   Hyderabad        5940  2000-01-01      23.779
        Hyderabad        5941  2000-02-01      25.826
        Hyderabad        5942  2000-03-01      28.821
        Hyderabad        5943  2000-04-01      32.698
        Hyderabad        5944  2000-05-01      32.438
...                       ...         ...         ...
Iraq    Baghdad          1150  2013-05-01      28.673
        Baghdad          1151  2013-06-01      33.803
        Baghdad          1152  2013-07-01      36.392
        Baghdad          1153  2013-08-01      35.463
        Baghdad          1154  2013-09-01         NaN

[2145 rows x 3 columns]
                          date  avg_temp_c
country     city                          
Afghanistan Kabul   2000-01-01       3.326
            Kabul   2000-02-01       3.454
            Kabul   2000-03-01       9.612
            Kabul   2000-04-01      17.925
            Kab

### Slicing time series

Slicing is particularly useful for time series since it's a common thing to want to filter for data within a date range. Add the `date` column to the index, then use `.loc[]` to perform the subsetting. The important thing to remember is to keep your dates in ISO 8601 format, that is, `"yyyy-mm-dd"` for year-month-day, `"yyyy-mm"` for year-month, and `"yyyy"` for year.

Recall from Chapter 1 that you can combine multiple Boolean conditions using logical operators, such as `&`. To do so in one line of code, you'll need to add parentheses `()` around each condition.

`pandas` is loaded as `pd` and `temperatures`, with no index, is available.

#### Instructions

* Use Boolean conditions, not `.isin()` or `.loc[]`, and the full date `"yyyy-mm-dd"`, to subset `temperatures` for rows where the `date` column is in 2010 and 2011 and print the results.
* Set the index of `temperatures` to the `date` column and sort it.
* Use `.loc[]` to subset `temperatures_ind` for rows in 2010 and 2011.
* Use `.loc[]` to subset `temperatures_ind` for rows from August 2010 to February 2011.


In [19]:
# Use Boolean conditions to subset temperatures for rows in 2010 and 2011
temperatures_bool = temperatures[( temperatures["date"] >= "2010-01-01") & (temperatures["date"] <= "2011-12-01")]
print(temperatures_bool)

# Set date as the index and sort the index
temperatures_ind = temperatures.set_index("date").sort_index()

# Use .loc[] to subset temperatures_ind for rows in 2010 and 2011
print(temperatures_ind.loc["2010": "2011"])

# Use .loc[] to subset temperatures_ind for rows from Aug 2010 to Feb 2011
print(temperatures_ind.loc["2010-08-01": "2011-02-01"])

       Unnamed: 0        date     city        country  avg_temp_c
120           120  2010-01-01  Abidjan  Côte D'Ivoire      28.270
121           121  2010-02-01  Abidjan  Côte D'Ivoire      29.262
122           122  2010-03-01  Abidjan  Côte D'Ivoire      29.596
123           123  2010-04-01  Abidjan  Côte D'Ivoire      29.068
124           124  2010-05-01  Abidjan  Côte D'Ivoire      28.258
...           ...         ...      ...            ...         ...
16474       16474  2011-08-01     Xian          China      23.069
16475       16475  2011-09-01     Xian          China      16.775
16476       16476  2011-10-01     Xian          China      12.587
16477       16477  2011-11-01     Xian          China       7.543
16478       16478  2011-12-01     Xian          China      -0.490

[2400 rows x 5 columns]
            Unnamed: 0        city    country  avg_temp_c
date                                                     
2010-01-01        4905  Faisalabad   Pakistan      11.810
2010-01-0

# Subsetting by row/column number

The most common ways to subset rows are the ways we've previously discussed: using a Boolean condition or by index labels. However, it is also occasionally useful to pass row numbers.

This is done using `.iloc[]`, and like `.loc[]`, it can take two arguments to let you subset by rows and columns.

`pandas` is loaded as `pd`. `temperatures` (without an index) is available.

## Instructions

Use `.iloc[]` on `temperatures` to take subsets.

* Get the 23rd row, 2nd column (index positions 22 and 1).
* Get the first 5 rows (index positions 0 to 5).
* Get all rows, columns 3 and 4 (index positions 2 to 4).
* Get the first 5 rows, columns 3 and 4.


In [21]:
# Get 23rd row, 2nd column (index 22, 1)
print(temperatures.iloc[22, 1])

# Use slicing to get the first 5 rows
print(temperatures.iloc[:5])

# Use slicing to get columns 3 to 4
print(temperatures.iloc[:, 2:4])

# Use slicing in both directions at once
print(temperatures.iloc[:5, 2:4])

2001-11-01
   Unnamed: 0        date     city        country  avg_temp_c
0           0  2000-01-01  Abidjan  Côte D'Ivoire      27.293
1           1  2000-02-01  Abidjan  Côte D'Ivoire      27.685
2           2  2000-03-01  Abidjan  Côte D'Ivoire      29.061
3           3  2000-04-01  Abidjan  Côte D'Ivoire      28.162
4           4  2000-05-01  Abidjan  Côte D'Ivoire      27.547
          city        country
0      Abidjan  Côte D'Ivoire
1      Abidjan  Côte D'Ivoire
2      Abidjan  Côte D'Ivoire
3      Abidjan  Côte D'Ivoire
4      Abidjan  Côte D'Ivoire
...        ...            ...
16495     Xian          China
16496     Xian          China
16497     Xian          China
16498     Xian          China
16499     Xian          China

[16500 rows x 2 columns]
      city        country
0  Abidjan  Côte D'Ivoire
1  Abidjan  Côte D'Ivoire
2  Abidjan  Côte D'Ivoire
3  Abidjan  Côte D'Ivoire
4  Abidjan  Côte D'Ivoire


## Working with pivot tables

### Pivoting the dog pack
```python
dogs_height_by_breed_vs_color = dog_pack.pivot_table("height_cm", index="breed", columns="color")

print(dogs_height_by_breed_vs_color)
```

The first argument is the column name containing values to aggregate. The index argument lists the columns to group by and display in rows, and the columns argument lists the columns to group by and display in columns


### `.loc[]` + `slicing` is a power combo

Pivot tables are just DataFrames with sorted indexes. That means that all the fun stuff you've learned so far this chapter can be used on them. In particular, the loc and slicing combination is ideal for subsetting pivot tables, like so.

```python
dogs_height_by_breed_vs_color.loc["Chow Chow":"Poodle"]
```

### The axis argument

The methods for calculating summary statistics on a DataFrame, such as mean, have an axis argument. The default value is "index," which means "calculate the statistic across rows." Here, the mean is calculated for each color. That is, "across the breeds." The behavior is the same as if you hadn't specified the axis argument.
```python
dogs_height_by_breed_vs_color.mean(axis="index")
```

### Calculating summary stats across columns
To calculate a summary statistic for each row, that is, "across the columns," you set axis to "columns." Here, the mean height is calculated for each breed. That is, "across the colors." For most DataFrames, setting the axis argument doesn't make any sense, since you'll have different data types in each column. Pivot tables are a special case since every column contains the same data type.

```python
dogs_height_by_breed_vs_color.mean(axis="columns")
```

### Pivot temperature by city and year

It's interesting to see how temperatures for each city change over time—looking at every month results in a big table, which can be tricky to reason about. Instead, let's look at how temperatures change by year.

You can access the components of a date (year, month and day) using code of the form `dataframe["column"].dt.component`. For example, the month component is `dataframe["column"].dt.month`, and the year component is `dataframe["column"].dt.year`.

Once you have the year column, you can create a pivot table with the data aggregated by city and year, which you'll explore in the coming exercises.

`pandas` is loaded as `pd`. `temperatures` is available.

#### Instructions

* Add a `year` column to `temperatures`, from the `year` component of the `date` column.
* Make a pivot table of the `avg_temp_c` column, with `country` and `city` as rows, and `year` as columns. Assign to `temp_by_country_city_vs_year`, and *look at the result* .


In [27]:
# Convert date column to datetime
temperatures["date"] = pd.to_datetime(temperatures["date"])
# Add a year column to temperatures
temperatures["year"] = temperatures["date"].dt.year

# Pivot avg_temp_c by country and city vs year
temp_by_country_city_vs_year = temperatures.pivot_table(values="avg_temp_c", index=["country", "city"], columns="year")

# See the result
print(temp_by_country_city_vs_year)

year                                 2000       2001       2002       2003  \
country       city                                                           
Afghanistan   Kabul             15.822667  15.847917  15.714583  15.132583   
Angola        Luanda            24.410333  24.427083  24.790917  24.867167   
Australia     Melbourne         14.320083  14.180000  14.075833  13.985583   
              Sydney            17.567417  17.854500  17.733833  17.592333   
Bangladesh    Dhaka             25.905250  25.931250  26.095000  25.927417   
...                                   ...        ...        ...        ...   
United States Chicago           11.089667  11.703083  11.532083  10.481583   
              Los Angeles       16.643333  16.466250  16.430250  16.944667   
              New York           9.969083  10.931000  11.252167   9.836000   
Vietnam       Ho Chi Minh City  27.588917  27.831750  28.064750  27.827667   
Zimbabwe      Harare            20.283667  20.861000  21.079333 

### Subsetting pivot tables

A pivot table is just a DataFrame with sorted indexes, so the techniques you have learned already can be used to subset them. In particular, the `.loc[]` + slicing combination is often helpful.

`pandas` is loaded as `pd`. `temp_by_country_city_vs_year` is available.

#### Instructions

Use `.loc[]` on `temp_by_country_city_vs_year` to take subsets.

* From Egypt to India.
* From Egypt, Cairo to India, Delhi.
* From Egypt, Cairo to India, Delhi, and 2005 to 2010.


In [30]:
# Subset for Egypt to India
print(temp_by_country_city_vs_year.loc["Egypt": "India"])

# Subset for Egypt, Cairo to India, Delhi
print(temp_by_country_city_vs_year.loc[("Egypt", "Cairo"):("India", "Delhi")])

# Subset for Egypt, Cairo to India, Delhi, and 2005 to 2010
print(temp_by_country_city_vs_year.loc[("Egypt", "Cairo"):("India", "Delhi"), "2005":"2010"])

year                       2000       2001       2002       2003       2004  \
country  city                                                                 
Egypt    Alexandria   20.744500  21.454583  21.456167  21.221417  21.064167   
         Cairo        21.486167  22.330833  22.414083  22.170500  22.081917   
         Gizeh        21.486167  22.330833  22.414083  22.170500  22.081917   
Ethiopia Addis Abeba  18.241250  18.296417  18.469750  18.320917  18.292750   
France   Paris        11.739667  11.371250  11.871333  11.909500  11.338833   
Germany  Berlin       10.963667   9.690250  10.264417  10.065750   9.822583   
India    Ahmadabad    27.436000  27.198083  27.719083  27.403833  27.628333   
         Bangalore    25.337917  25.528167  25.755333  25.924750  25.252083   
         Bombay       27.203667  27.243667  27.628667  27.578417  27.318750   
         Calcutta     26.491333  26.515167  26.703917  26.561333  26.634333   
         Delhi        26.048333  25.862917  26.63433

### Calculating on a pivot table

Pivot tables are filled with summary statistics, but they are only a first step to finding something insightful. Often you'll need to perform further calculations on them. A common thing to do is to find the rows or columns where the highest or lowest value occurs.

Recall from Chapter 1 that you can easily subset a Series or DataFrame to find rows of interest using a logical condition inside of square brackets. For example: `series[series > value]`.

`pandas` is loaded as `pd` and the DataFrame `temp_by_country_city_vs_year` is available. The `.head()` for this DataFrame is shown below, with only a few of the year columns displayed:


| country     | city      | 2000   | 2001   | 2002   | … | 2013   |
| ------------- | ----------- | -------- | -------- | -------- | ---- | -------- |
| Afghanistan | Kabul     | 15.823 | 15.848 | 15.715 | … | 16.206 |
| Angola      | Luanda    | 24.410 | 24.427 | 24.791 | … | 24.554 |
| Australia   | Melbourne | 14.320 | 14.180 | 14.076 | … | 14.742 |
|             | Sydney    | 17.567 | 17.854 | 17.734 | … | 18.090 |
| Bangladesh  | Dhaka     | 25.905 | 25.931 | 26.095 | … | 26.587 |

#### Instructions

* Calculate the mean temperature for each year, assigning to `mean_temp_by_year`.
* Filter `mean_temp_by_year` for the year that had the highest mean temperature.
* Calculate the mean temperature for each city (across columns), assigning to `mean_temp_by_city`.
* Filter `mean_temp_by_city` for the city that had the lowest mean temperature.


In [40]:
# Get the worldwide mean temp by year
mean_temp_by_year = temp_by_country_city_vs_year.mean()

# Filter for the year that had the highest mean temp
print(mean_temp_by_year[mean_temp_by_year == mean_temp_by_year.max()])

# Get the mean temp by city
mean_temp_by_city = temp_by_country_city_vs_year.mean(axis="columns")

# Filter for the city that had the lowest mean temp
print(mean_temp_by_city[mean_temp_by_city == mean_temp_by_city.min()])

year
2013    20.312285
dtype: float64
country  city  
China    Harbin    4.876551
dtype: float64
